File: frontend-web/types/aggregator.ts
export interface AggregatorHolding {
  id: string; // assignment id
  artworkId: string;
  advancePercent: 5 | 3;
  advanceAmount: number;
  // Paid with the advance before taking possession (MOU §7). The money-flow
  // sheet returns it only if the piece sells — an unsold piece going back to
  // GalleryZone refunds the advance alone. Optional because fixture holdings
  // predate the field.
  deliveryDeposit?: number;
  displayPrice: number; // aggregator-editable; floor = artwork.customerPrice
  assignedAt: string; // ISO
  expiresAt: string; // assignedAt + 30 days
  // "returned" is kept rather than deleted: the artwork's cycle month is
  // counted from how many aggregators have already had it, so the history is
  // what decides the next aggregator's price and advance.
  status: "reserved" | "sold_pending_settlement" | "returned";
  // Aggregator MOU §6: the aggregator gets ONE opportunity to set the selling
  // price. Stamped the first time they set it; after that the price is locked.
  // Optional because fixture holdings predate the field — undefined and null
  // both mean "not set yet".
  displayPriceSetAt?: string | null;
  // Which month of the artwork's five-month aggregator cycle this placement is.
  // A piece that doesn't sell moves to a DIFFERENT aggregator each month, at a
  // lower price and a different advance rate, so the month is a property of the
  // artwork's journey rather than of any one aggregator. Optional because
  // fixture holdings predate the field; absent means month 1.
  cycleMonth?: number;
  /** Set when the piece went back to GalleryZone unsold. */
  returnedAt?: string | null;
  // True when this placement runs past the usual thirty days because what would
  // have been left of the artist's 180 days was too short to hand to anyone
  // else. The last aggregator keeps it rather than the piece making one more
  // journey for a fortnight.
  windowExtended?: boolean;
  // Distinguishes an aggregator-initiated reservation (Browse → Reserve) from
  // GalleryZone proactively placing inventory at this aggregator's premises
  // (MOU §4). Drives the My Inventory filter tabs.
  assignmentSource: "self_reserved" | "gz_assigned";
}

export interface RecordSalePayload {
  artworkId: string;
  soldPrice: number;
  buyerName: string;
  buyerEmail: string;
  buyerPhone: string;
  deliveryAddress: {
    line1: string;
    city: string;
    state: string;
    pincode: string;
  };
  deliveryMode: "courier" | "self_pickup";
  paymentRoute: "direct_to_galleryzone" | "cash_at_premises";
}

// Persisted result of recordSale() — buyer/price/delivery details Orders &
// Sales, Customers, Shipping, Wallet crediting, and Settlements need.
export interface AggregatorSale {
  id: string;
  holdingId: string;
  artworkId: string;
  soldPrice: number;
  buyerName: string;
  buyerEmail: string;
  buyerPhone: string;
  deliveryAddress: {
    line1: string;
    city: string;
    state: string;
    pincode: string;
  };
  deliveryMode: "courier" | "self_pickup";
  // The aggregator collects "on behalf of GalleryZone", never for themselves.
  // Either the buyer paid GalleryZone directly (transfer or UPI, using the
  // details on the checkout page), or the aggregator took cash — in which case
  // they owe GalleryZone the WHOLE sale price and their commission is settled
  // separately afterwards. They never net it off at the counter.
  paymentRoute?: "direct_to_galleryzone" | "cash_at_premises";
  /** Set when the aggregator has transferred cash they collected. */
  remittedAt?: string | null;
  soldAt: string; // ISO
  shipmentStatus: "preparing" | "dispatched" | "delivered";
  dispatchedAt: string | null;
  deliveredAt: string | null;
  courierRef: string | null; // null when deliveryMode === "self_pickup"
}

// The aggregator's own physical premises (not the Artist Dashboard's
// "Gallery Spaces" page, which lists artworks placed at an aggregator).
export interface GallerySpace {
  id: string;
  name: string;
  addressLine1: string;
  city: string;
  state: string;
  pincode: string;
  capacity: number; // max pieces this location can display at once
  coordinatorName: string; // MOU §10: "nominate one Galleryzone coordinator"
}

File: mobile_flutter/lib/data/models/aggregator.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'artist_portal.dart';
import 'artwork.dart';

part 'aggregator.freezed.dart';
part 'aggregator.g.dart';

/// Aggregator-portal shapes: `types/aggregator.ts` plus the aggregator's own
/// profile/settings collections in `lib/mock-collections.ts`.
///
/// [AggregatorHolding] itself already lives in `artist_portal.dart` — the
/// artist side reads holdings (Gallery Spaces), this side writes them.

enum DeliveryMode {
  courier,
  @JsonValue('self_pickup')
  selfPickup,
}

enum ShipmentStatus { preparing, dispatched, delivered }

/// The aggregator collects "on behalf of GalleryZone", never for themselves.
/// Either the buyer paid GalleryZone directly (transfer or UPI, using the
/// details on the checkout screen), or the aggregator took cash — in which
/// case they owe GalleryZone the WHOLE sale price and their commission is
/// settled separately afterwards. They never net it off at the counter.
enum PaymentRoute {
  @JsonValue('direct_to_galleryzone')
  directToGalleryZone,
  @JsonValue('cash_at_premises')
  cashAtPremises,
}

const paymentRouteLabel = {
  PaymentRoute.directToGalleryZone: 'Buyer paid GalleryZone directly',
  PaymentRoute.cashAtPremises: 'Cash taken at the gallery',
};

/// The buyer's address as the aggregator records it at the counter. Distinct
/// from `order.dart`'s `Address`, which is a saved, reusable address-book
/// entry with an id and a default flag — this one is captured once, on one
/// sale, and never reused.
@freezed
abstract class DeliveryAddress with _$DeliveryAddress {
  const factory DeliveryAddress({
    required String line1,
    required String city,
    required String state,
    required String pincode,
  }) = _DeliveryAddress;

  factory DeliveryAddress.fromJson(Map<String, dynamic> json) =>
      _$DeliveryAddressFromJson(json);
}

/// Persisted result of `recordSale()` — the buyer/price/delivery detail the
/// Orders, Customers, Shipping, Wallet and Settlements screens all read.
@freezed
abstract class AggregatorSale with _$AggregatorSale {
  const factory AggregatorSale({
    required String id,
    required String holdingId,
    required String artworkId,
    required double soldPrice,
    required String buyerName,
    required String buyerEmail,
    required String buyerPhone,
    required DeliveryAddress deliveryAddress,
    required DeliveryMode deliveryMode,
    required String soldAt,
    required ShipmentStatus shipmentStatus,
    String? dispatchedAt,
    String? deliveredAt,

    /// How the buyer's money reached GalleryZone.
    @Default(PaymentRoute.directToGalleryZone) PaymentRoute paymentRoute,

    /// Set when the aggregator has transferred cash they collected. Null
    /// while the money is still sitting in their till.
    String? remittedAt,

    /// Null when [deliveryMode] is [DeliveryMode.selfPickup].
    String? courierRef,
  }) = _AggregatorSale;

  factory AggregatorSale.fromJson(Map<String, dynamic> json) =>
      _$AggregatorSaleFromJson(json);
}

/// The aggregator's own physical premises. Not the artist portal's "Gallery
/// Spaces" screen, which lists artworks placed *at* an aggregator — same
/// label, opposite direction, different actor.
@freezed
abstract class GallerySpace with _$GallerySpace {
  const factory GallerySpace({
    required String id,
    required String name,
    required String addressLine1,
    required String city,
    required String state,
    required String pincode,

    /// Max pieces this location can display at once.
    required int capacity,

    /// MOU §10: "nominate one Galleryzone coordinator".
    required String coordinatorName,
  }) = _GallerySpace;

  factory GallerySpace.fromJson(Map<String, dynamic> json) =>
      _$GallerySpaceFromJson(json);
}

@freezed
abstract class AggregatorProfile with _$AggregatorProfile {
  const factory AggregatorProfile({
    required String companyName,
    required String contactPerson,
    required String avatar,
    required String gstNumber,
    required String phone,
    required String addressLine1,
    required String bankAccountMasked,
    required String ifsc,
    required String securityDepositStatus,

    /// Signed once from the aggregator's profile screen — the partner
    /// agreement, separate from the artist MOU. Null until they sign it, and
    /// reserving is refused until they do: an unsigned aggregator has no
    /// agreement covering custody, pricing or settlement, so they cannot take
    /// possession of anyone's artwork.
    MouAcceptance? mouAcceptance,
  }) = _AggregatorProfile;

  factory AggregatorProfile.fromJson(Map<String, dynamic> json) =>
      _$AggregatorProfileFromJson(json);
}

@freezed
abstract class AggregatorSettings with _$AggregatorSettings {
  const factory AggregatorSettings({
    required bool notifyNewAssignment,
    required bool notifySaleRecorded,
    required bool notifySettlementProcessed,
    required bool notifyExpiryReminder,
  }) = _AggregatorSettings;

  factory AggregatorSettings.fromJson(Map<String, dynamic> json) =>
      _$AggregatorSettingsFromJson(json);
}

/// A holding joined to the artwork it points at. Derived on read, never
/// persisted — the holding record itself carries only an `artworkId`.
class AggregatorHoldingView {
  const AggregatorHoldingView({required this.holding, required this.artwork});

  final AggregatorHolding holding;
  final Artwork artwork;
}

/// Buyers rolled up from recorded sales, keyed by email. Derived, not a
/// stored customer record — this portal never owns a customer table.
class AggregatorCustomer {
  const AggregatorCustomer({
    required this.buyerName,
    required this.buyerEmail,
    required this.buyerPhone,
    required this.orderCount,
    required this.totalSpend,
  });

  final String buyerName;
  final String buyerEmail;
  final String buyerPhone;
  final int orderCount;
  final double totalSpend;
}

/// The three dashboard headline figures.
class AggregatorDashboardSummary {
  const AggregatorDashboardSummary({
    required this.activeReservations,
    required this.commissionEarned,
    required this.pendingSettlements,
  });

  final int activeReservations;
  final double commissionEarned;
  final int pendingSettlements;
}

class AggregatorAnalyticsSummary {
  const AggregatorAnalyticsSummary({
    required this.salesCount,
    required this.totalRevenue,
    required this.commissionPending,
    required this.commissionAvailable,
    required this.customerCount,
    required this.activeReservations,
    required this.averageSoldPrice,
    required this.averageDisplayMarkup,
  });

  final int salesCount;
  final double totalRevenue;
  final double commissionPending;
  final double commissionAvailable;
  final int customerCount;
  final int activeReservations;
  final double averageSoldPrice;
  final double averageDisplayMarkup;
}

/// Revenue by artwork category, derived from recorded sales.
class CategoryPerformance {
  const CategoryPerformance({
    required this.category,
    required this.revenue,
    required this.orders,
  });

  final String category;
  final double revenue;
  final int orders;
}
